week2/homework/exercises/1-blog-API/server.js

```diff
@@ -1,10 +1,79 @@
+const fs = require("fs"); 
 const express = require('express')
 const app = express();
- 
+app.use(express.json());
 
 // YOUR CODE GOES IN HERE
-app.get('/', function (req, res) {
-  res.send('Hello World')
+// GET - READ /blogs
+// POST - CREATE /blogs
+// UPDATE - PUT /posts/:id
+// DELETE - DELETE /blogs/:id
+// app.get('/', function (req, res) {
+//   res.send('Hello World')
+// })
+// Creating new posts
+app.post('/blogs', (req, res) => {
+  let title = req.body.title;
+  let content = req.body.content;
+  res.writeHead(200, "OK", {'Content-Type': 'application/json'});
+  fs.writeFileSync(title, content);
+  res.end('ok')
 })
- 
-app.listen(3000)
\ No newline at end of file
+// Updating existing posts
+app.put('/posts/:title', (req, res) => {
+  // How to get the title and content from the request?
+  let title = req.params.title;
+  let content = req.body.content;
+  if (fs.existsSync(title)) {
+    res.writeHead(200, "OK", {'Content-Type': 'application/json'});
+    fs.writeFileSync(title, content);
+    res.end('ok')
+  }
+  else {
+    // Send response with error message
+    res.status(400).end('This post does not exist!');
+  }
+})
+// Deleting posts
+app.delete('/blogs/:title', (req, res) => {
+  // How to get the title from the url parameters?
+  let title = req.params.title;
+  if (fs.existsSync(title)) { 
+    fs.unlinkSync(title);
+    res.end('ok');
+  } else {
+    // Respond with message here
+    res.status(400).end('This post does not exist!');
+  }
+})
+// Reading posts
+app.get('/blogs/:title', (req, res) => {
+  // How to get the title from the url parameters?
+  let title = req.params.title;
+  // check if post exists
+  if (fs.existsSync(title)) { 
+    const post = fs.readFileSync(title);
+    res.end(post);
+  } else {
+    res.status(400).end('This post does not exist!');
+  }
+})
+
+// bonus
+app.get('/blogs', (req, res) => {
+  // how to get the file names of all files in a folder??
+  const folder = './blogs';
+  fs.readdir(folder, { withFileTypes: true }, (err, files) => {
+    let title = req.params.title;
+    if (fs.existsSync(folder)) { 
+      files = files.map(file => {
+        return file;
+      }); 
+      res.end(JSON.stringify(files));
+    } else {
+      res.status(400).end(err);
+    }
+  })
+})
```

Comment:
You were doing so good and messed it up a bit in this request handler.
Here are some suggestions  
1. no need to use async calls when using sync calls everywhere else  
2. req.params.title is not part of this call
3. checking if the folder exists after reading its contents does not make a lot of sense
4. if the folder does not exist its safe to assume that there are no blogs so we can send an empty array as a reponse instead of an error

```suggestion
  const folder = './blogs';
  if ( ! fs.existsSync(folder)) { 
      res.end([]); // send back an empty array
      return res;
  }
  fs.readdir(folder, { withFileTypes: true }, (err, files) => {
      if ( err ) {
         console.error(err);
         res.status(500).end('Something went wrong');
      } else {
         let filenames = files.map(file => file.name); // ?? not sure from the top of my head if name is the right property
         res.end(JSON.stringify(files));
      }
  })
```

Reply:
Thank you, Andrej! I got it. I will fix my code according to your notice.